### PR TITLE
Higher-level Pebble integration into Operator Framework

### PR DESCRIPTION
**Pull request by benhoyt**
Includes addition of container-related metadata parsing (this is under discussion and may change!) Also rename `pebble.API` to `pebble.PebbleClient` per discussion. Still need to add tests - doing that now.

Initial work / feedback was done on [benhoyt/operator PR #1](https://github.com/benhoyt/operator/pull/1) ... this is squashed and rebased onto master now that the initial pebble PR is merged.

Intended usage is something like this:

```python
container = model.unit.containers['pgweb']
container = model.unit.get_container('pgweb')  # same as above

container.add_layer("""
summary: Simple override layer
services:
    srv1:
        override: merge
        environment:
            - var1: val1
""")
container.add_layer({  # supplying a dict works too
    'summary': 'Another layer',
    'services': {
        'srv1': {
            'override': 'merge',
            'environment': my_environ,
        },
    },
})
flattened_setup = container.get_setup()  # rich pebble.Setup object
print(flattened_setup.to_yaml())

container.start('foo', 'bar')
container.stop('foo')
container.stop('bar')
container.autostart()
```

Or you can drop down to manually accessing the `pebble.API` instance if you want:

```python
pebble = model.containers['pgweb'].pebble
pebble.stop_services(['postgres', 'redis'])
do_backup()
pebble.start_services(['postgres', 'redis'])
```

Notes:

* There can be multiple workload containers in a single charm, so need to distinguish them by name.
  - see https://github.com/hpidcock/testing-charms/blob/master/pebbletest/metadata.yaml
  - hence the need for model.get_container(name) instead of just model.container
* Unit ~= Pod in K8s

### CODE REVIEW

File: ops/model.py
Comment (niemeyer) [inline ops/model.py:278]: If we do that it means we cannot ever make this not a dictionary, as it will break compatibility with existing clients. I think we've used that rationale to move them towards a richer type, which enables us to implement richer functionality around the type (such as lazy loading).

Should we do the same here?

File: ops/model.py
Comment (niemeyer) [inline ops/model.py:1039]: Is there a convention in place for this already?  I remember we discussed this during one of our calls, and it felt a bit awkward that we say "Attributes", but "Args". If we still have a choice, I'd spell it out as "Arguments:" or "Parameters:' here as well, or go the opposite way and shorten them all.

File: ops/model.py
Comment (niemeyer) [inline ops/model.py:1037]: Either "... add a Pebble setup layer.", or just "... add a setup layer." might feel a bit better I think.

File: ops/model.py
Comment (niemeyer) [inline ops/model.py:1041]: I now realize there's a small conceptual clash here.  In Pebble, the Setup type contains the entire Pebble setup, with all the layers and any additional global settings relevant to the entire process. Here, we're talking about a Layer, not a complete Setup.  I believe the Pebble code remains true to this idea, or is there an inconsistency on the other side as well?

File: ops/model.py
Comment (niemeyer) [inline ops/model.py:1048]: There are at least two ways out of this, I think: either we make this a layer API, or we turn setup into something that owns a list of layers which happens to have a single layer in this case. What do you think?

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:356]: That would sound better as PebbleService, I think, or pebble.Service, if we're buying into that convention (pebble.Setup or pebble.Layer probably means we're already there).

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:370]: What's the borrowing semantics we intend for this type?  Do we care? That is, if people provide a raw value in, are we a) copying it; or b) taking it over and potentially allowing further code to mutate it?

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:315]: Perhaps something resembling:
```
"""Initialize client instance.

Arguments:

    socket_path: .... - defaults to $PEBBLE/.pebble.socket.

...
"""
```

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:403]: Logic above doesn't seem to reflect the documentation. opener and socket_path both default to None, so it will error out instead of assuming default values.

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:531]: dict?

**Review by niemeyer — CHANGES_REQUESTED** [review]
Thanks for the tweaks, Ben. Here is another pass for us to discuss. Seems pretty much there.

File: ops/model.py
Comment (benhoyt) [reply ops/model.py]: Fixed. I've added a `ContainerMapping` type for this. Thought about using the existing `LazyMapping`, but it's not really lazy so just added a simple dedicated type. Also means it's not mutable, which I think is a good thing.

File: ops/model.py
Comment (benhoyt) [reply ops/model.py:1039]: Yeah, `Args:` is used everywhere throughout the project. We're using the "google" style for our docstrings, which uses `Attributes:` and `Args:` for whatever reason: https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html

File: ops/model.py
Comment (benhoyt) [reply ops/model.py]: Good point -- a setup has a list of layers. I've changed to just `pebble.Layer` as that's all that we need here. And `get_setup` returns a flattened setup, which essentially means flattened to a single layer.

File: ops/model.py
Comment (benhoyt) [reply ops/model.py]: Per above, I've gone the `pebble.Layer` route as I think that's what we need here.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Changed to `pebble.Service`.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Good call. I've changed these (also the lists) to make copies and added a test for that.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py:315]: I think you're looking at the deleted code -- I've actually removed the environment variable default and the env var handling from the `PebbleClient` class, and only use that in the test command line client. In `model.py` the socket_path arg is always passed in.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py:403]: See above -- I had updated this. Now says:

> Defaults to using a Unix socket at socket_path (which must be specified
> unless a custom opener is provided).

File: ops/pebble.py
Comment (niemeyer) [reply ops/pebble.py:315]: Duh, thanks.

File: ops/pebble.py
Comment (niemeyer) [reply ops/pebble.py:403]: We should probably make it behave similar to the pebble client looking at the environment, but there's nothing pressing it to be done right now.

**Review by niemeyer — APPROVED** [review]
Thanks for the changes. One last review, and it LGTM once these issues have been addressed.

File: ops/model.py
Comment (benhoyt) [reply ops/model.py]: No reason, except "it seemed better to append the new parameter rather than insert". But this is a purely internal class so makes no difference -- fixed.

File: ops/model.py
Comment (benhoyt) [reply ops/model.py]: Per discussion, changed this to `get_layer`.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Thanks -- done.

File: test/fake_pebble.py
Comment (jameinel) [inline test/fake_pebble.py:26]: I didn't get the time to review the Handler yet.

File: test/test_model.py
Comment (jameinel) [inline test/test_model.py:792]: It feels a little weird to be mocking one dependency, but passing in the others. It might still be the correct way to do it, though.

File: test/test_model.py
Comment (jameinel) [inline test/test_model.py:839]: import mock

is part of the stdlib, and helps avoid some of the boilerplate here, but also tends to have a little bit too much magic. Mock with spec_set tends to avoid some of the obvious pitfalls (it makes sure the signatures of the methods are correct, rather than just accepting anything)
Something to consider, at least.

File: test/test_pebble.py
Comment (jameinel) [inline test/test_pebble.py:704]: Does pebble return a status: OK for an error response?

File: test/test_pebble.py
Comment (jameinel) [inline test/test_pebble.py:742]: IRC, there was some issue about platform and using win32, but when I dug the only thing I found was that linux2 was renamed to linux back in python 3.3 (because Linus changed something that would have made it linux3 according to the kernel version.)
I did confirm that python3.7 in 64 bit mode does, indeed, report win32.

**Review by jameinel — COMMENTED** [review]
i'd like to get a chance to read over the fake Pebble, before giving a final response, but it is looking quite good.

File: ops/model.py
Comment (benhoyt) [reply ops/model.py]: Thanks. Per discussion, I now explicitly pass `meta` to each cache entity class's `__init__` method (e.g., `Unit` and `Application`), like we do with backend. This definitely feels better than using `cache._meta` from somewhere else, and it was a pretty small change.

File: ops/model.py
Comment (benhoyt) [reply ops/model.py:355]: Yeah, this was per discussion with Gustavo et al. A little weird, but as you say, the `get_container` allows more functionality to be added later.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Thanks! Switched the use the fast version if available. (I also noticed a couple of other places in our code where we weren't using the fast version.)

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Done -- added `ops/_yaml.py` to unify this. We were defining it in three places, so made sense. Even though it's fairly trivial, it was also messy.

File: test/test_model.py
Comment (benhoyt) [reply test/test_model.py:839]: I've given this a look, and I guess I don't love `Mock`, as it requires you to learn a whole new DSL instead of just writing a few lines of clear Python. We're not using it elsewhere in the project, so there's not a precedent. Happy to switch in another PR if we really decide to go all in on `Mock`.

File: test/test_pebble.py
Comment (benhoyt) [reply test/test_pebble.py:704]: Yeah, the overall request succeeded, even though one of the operations had an error. This makes a bit more sense when there are multiple changes/services at play -- for example, if you call `start('foo', 'bar')` and service `foo` starts but `bar` doesn't, overall it returns success even though one of the changes had an error.

File: test/test_helpers.py
Comment (jameinel) [inline test/test_helpers.py:150]: The only test I would add is that the load and dump are genuinely safe and refuse non basic content. Serializing a class should should sufficient to trigger an exception. I don't remember the exact yaml syntax but you can serialize with regular unsafe dump to get the content that would be unsafe to load.

**Review by jameinel — APPROVED** [review]

File: ops/model.py
Comment (benhoyt) [reply ops/model.py]: Done

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py:400]: Yep -- the testing Python CLI does that, but not at the `pebble.Client` level, at the CLI/main level.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Added

File: test/fake_pebble.py
Comment (benhoyt) [reply test/fake_pebble.py:26]: No worries -- can always adjust in a follow-up PR if you have further feedback.

File: test/test_model.py
Comment (benhoyt) [reply test/test_model.py:792]: This seemed to me the simplest way to write this test, so keeping as is for now.

File: test/test_pebble.py
Comment (benhoyt) [reply test/test_pebble.py:742]: Yep, that's my take too -- `win32` seems to be the documented, correct value for `sys.platform` on Windows, whether 64-bit or not.

File: test/test_helpers.py
Comment (benhoyt) [reply test/test_helpers.py:160]: Agreed, thanks -- done.

File: ops/model.py
Comment (niemeyer) [inline ops/model.py:1019]: The name doesn't seem enough to determine which pebble is to be used on the local system. It is assuming that it is a container, but pebble doesn't require containers to actually work.  It need to either take a socket, or something else that enables us to distinguish on the backend side what to do with that name.

File: ops/_yaml.py
Comment (niemeyer) [inline ops/_yaml.py:25]: Why are we not using yaml.safe_load and yaml.safe_dump themselves?

Can we do that and remove this logic and file entirely?

**Review by niemeyer — CHANGES_REQUESTED** [review]
Thanks for the changes Ben.  I'm merging this PR as it's blocked on Travis, but there are a couple of open points about changes that were made since I reviewed it.  Can you please have a look at these and let me know what you think?

File: ops/_yaml.py
Comment (benhoyt) [reply ops/_yaml.py:25]: That would be correct, but slow, because `safe_load` and `safe_dump` always use the slow Python versions of the loader, `SafeLoader`, whereas this uses `CSafeLoader` if it's available. See John's performance numbers above -- it makes quite a difference (not that performance is critical here). It's kinda daft that this logic isn't baked into PyYAML itself (maybe they didn't because CSafeLoader isn't quite backwards compatible with SafeLoader, I dunno?).

File: ops/model.py
Comment (benhoyt) [reply ops/model.py:1019]: Good call. Fix in https://github.com/canonical/operator/pull/473 -- just moving the container-specific building of `socket_path` from `_ModelBackend.get_pebble` to `Container.__init__`.

File: ops/_yaml.py
Comment (niemeyer) [reply ops/_yaml.py:25]: For the record, we discussed this online and got to an agreement of keeping it as a file but moving it to a _private directory.

File: ops/model.py
Comment (niemeyer) [reply ops/model.py:1019]: Thanks!

**Pull request merged**
